### PR TITLE
feat: support GDPO (New)

### PR DESCRIPTION
**Pull request by nbasyl**
# What does this PR do ?

**Add a one line overview of what this PR aims to accomplish.**

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

## Summary by CodeRabbit

* **New Features**
  * Added GDPO algorithm support for model training.
  * Integrated GSM8K dataset for mathematics problem training.
  * Enabled multi-reward training with separate reward components.
  * Added math environment supporting multiple distinct reward signals.

* **Tests**
  * Added GDPO functional test suite.

### CODE REVIEW

File: nemo_rl/algorithms/advantage_estimator.py [hunk 1]
```diff
@@ -37,14 +42,21 @@ def __init__(self, estimator_config: dict, loss_config: dict):
         self.use_leave_one_out_baseline = estimator_config["use_leave_one_out_baseline"]
         self.normalize_rewards = estimator_config["normalize_rewards"]
 
-    def compute_advantage(self, prompt_ids, rewards, mask, **kwargs):
+    def compute_advantage(
```
Comment (yuki-97) [inline nemo_rl/algorithms/advantage_estimator.py]: can we not change the code of `GRPOAdvantageEstimator` and `ReinforcePlusPlusAdvantageEstimator` to keep them the same as before?

File: tests/unit/algorithms/test_grpo.py [hunk 2]
```diff
@@ -1642,8 +1642,14 @@ def test_grpo_advantage_estimator_zero_std():
...
     mask = torch.ones(4, 5)
+    repeated_batch = {"_input_ids_for_baseline": prompt_ids, "total_reward": rewards}
 
-    result = estimator.compute_advantage(prompt_ids, rewards, mask)
+    result = estimator.compute_advantage(
+        prompt_ids=prompt_ids,
+        rewards=rewards,
+        repeated_batch=repeated_batch,
+        mask=mask,
+    )
```
Comment (yuki-97) [inline tests/unit/algorithms/test_grpo.py]: can you cleanup `_input_ids_for_baseline` in this file?
```suggestion

    result = estimator.compute_advantage(
        prompt_ids=prompt_ids,
        rewards=rewards,
        mask=mask,
    )
```

File: nemo_rl/experience/rollouts.py [hunk 3]
```diff
@@ -327,7 +327,13 @@ def calculate_rewards(
...
         range(len(all_indices_order)), key=lambda k: all_indices_order[k]
     )
-    rewards = torch.tensor([all_rewards[i] for i in sorted_indices])
+    # Stack rewards: each element may be scalar (single-reward env) or 1d (multi-reward env).
+    # Envs may return Python floats or tensors; ensure tensors for torch.stack.
+    # Handle empty batch (torch.stack requires non-empty list).
+    if not sorted_indices:
+        rewards = torch.tensor([], dtype=torch.float32)
+    else:
+        rewards = torch.stack([torch.as_tensor(all_rewards[i]) for i in sorted_indices])
```
Comment (yuki-97) [inline nemo_rl/experience/rollouts.py]: can we keep the original code and use if/else to handle with multi-reward?
```suggestion
    # Stack rewards: each element may be scalar (single-reward env) or 1d (multi-reward env).
    if len(all_rewards) > 0 and isinstance(all_rewards[0], torch.Tensor):
        rewards = torch.stack([all_rewards[i] for i in sorted_indices])
    else:
        rewards = torch.tensor([all_rewards[i] for i in sorted_indices])
```

File: tests/unit/algorithms/test_grpo.py [hunk 4]
```diff
@@ -1642,8 +1642,14 @@ def test_grpo_advantage_estimator_zero_std():
         [2.0, 2.0, 1.0, 3.0]
     )  # prompt 0: std=0; prompt 1: std=sqrt(2)
     mask = torch.ones(4, 5)
+    repeated_batch = {"_input_ids_for_baseline": prompt_ids, "total_reward": rewards}
 
-    result = estimator.compute_advantage(prompt_ids, rewards, mask)
+    result = estimator.compute_advantage(
```
Comment (yuki-97) [inline tests/unit/algorithms/test_grpo.py:1659]: can you add an unit test for GDPO adv estimator? you can refer to `test_reinforce_plus_plus_global_normalization` in this file.

File: nemo_rl/algorithms/grpo.py [hunk 5]
```diff
@@ -1055,7 +1062,14 @@ def _create_advantage_estimator(master_config: MasterConfig):
     )
 
     adv_estimator_name = adv_estimator_config["name"]
-    if adv_estimator_name == "grpo":
+    if adv_estimator_name == "gdpo":
+        assert not _should_use_async_rollouts(master_config), (
+            "GDPO is not supported for async rollouts, "
+            "please set policy.generation.vllm_cfg.async_engine to false in your config."
```
Comment (terrykong) [inline nemo_rl/algorithms/grpo.py]: can you leave a github issue link here about this support being missing, to use this for nemotron, we'll need this supported eventually so the github issue can serve as a TODO

File: examples/run_grpo.py [hunk 6]
```diff
@@ -139,6 +139,13 @@ def main() -> None:
                 "use_multiple_dataloader is not supported with async GRPO"
             )
 
+        # Async GDPO is not supported
+        if config["grpo"]["adv_estimator"]["name"] == "gdpo":
+            raise NotImplementedError(
+                "GDPO is not supported for async training, "
+                "please set grpo.async_grpo.enabled to false in your config."
```
Comment (terrykong) [inline examples/run_grpo.py]: same comment about a github issue here too

File: nemo_rl/environments/interfaces.py [hunk 7]
```diff
@@ -44,7 +44,7 @@ class EnvironmentReturn(NamedTuple, Generic[MetadataT]):
     observations: list[dict[str, str]]
     metadata: list[MetadataT]
     next_stop_strings: list[list[str] | None] | list[None]
-    rewards: Tensor
+    rewards: Tensor  ## This could be of different shape
```
Comment (terrykong) [inline nemo_rl/environments/interfaces.py]: nit: could this comment be a little more specific about the shapes? something like [B] | [B,num_reward]  or however the shape is arranged?

File: nemo_rl/environments/math_environment.py [hunk 8]
```diff
@@ -263,6 +365,56 @@ def shutdown(self) -> None:
...
+        Every rank will run this function, so you're free to use distributed
+        calculations if you'd prefer for heavy metrics.
+        """
+        # for multi-reward environment, index 0 always store corretness reward
+        rewards = (
+            batch["rewards"] if batch["rewards"].ndim == 1 else batch["rewards"][:, 0]
+        )
+
+        # set a reward of 0 for any incorrectly ended sequences
+        rewards = rewards * batch["is_end"]
```
Comment (terrykong) [inline nemo_rl/environments/math_environment.py:382]: it looks like there's a subtle behavior change here. we used to mask the rewards in place before:

```
batch["rewards"] = (
            batch["rewards"] * batch["is_end"]
```

could you elaborate on why that's not needed anymore?

File: nemo_rl/experience/rollouts.py [hunk 9]
```diff
@@ -536,6 +564,11 @@ def run_multi_turn_rollout(
     # Add total rewards to the final batch
     current_batch["total_reward"] = total_rewards
     current_batch["truncated"] = sample_truncated
+    # Expose per-component rewards (reward1, reward2, ...) for multi-reward envs only; GRPO uses total_reward
+    if multi_rewards is not None:
+        num_reward_components = multi_rewards.shape[1]
+        for i in range(num_reward_components):
+            current_batch[f"reward{i + 1}"] = multi_rewards[:, i].clone()
```
Comment (terrykong) [inline nemo_rl/experience/rollouts.py:571]: in a multi-environment rollout, won't we have issues with this if one environment has [math_format, math_correctness] as rewards but then has [code_correctness, code_format] so different orderings?

should environments then return scalar rewards or a dict of rewards so we can figure this out from the metrics? otherwise it looks like we need to look at code to figure out which reward is which, also multiple rewards of different types may get combined in unexpected way if they are positional

File: nemo_rl/environments/math_environment.py [hunk 10]
```diff
@@ -323,12 +475,11 @@ def step(
         )
 
         for worker_result in worker_results:
+            worker_scores = worker_result
             if return_extracted_answer:
                 worker_scores, worker_answers = worker_result
-                results.extend(worker_scores)
```
Comment (terrykong) [inline nemo_rl/environments/math_environment.py:328]: this looks different than before. is worker_result no longer a tuple of scores/answers? will this break eval @yuki-97?

File: nemo_rl/environments/math_environment.py (hunk 8 above)
Comment (yuki-97) [reply nemo_rl/environments/math_environment.py:382]: good catch for the in place change!

hmm, but I took a look at the code, seems `global_post_process_and_metrics` isn't actually used anywhere, so I think this change should be fine.
we should cleanup or make use of this function in another PR. correct me if I missed anything.

File: nemo_rl/environments/math_environment.py (hunk 10 above)
Comment (yuki-97) [reply nemo_rl/environments/math_environment.py:328]: the logic here should be the same as before, `worker_result` is still a tuple of scores/answers if `return_extracted_answer` is True, so I think here won't break eval.

but actually eval will be broken if we use the new added multi-reward math env (multi-reward is not adapted), so it should be good to add an assert to guard not using the env at eval for now. also good to file an issue to track. @nbasyl

File: nemo_rl/algorithms/advantage_estimator.py (hunk 1 above)
Comment (yuki-97) [reply nemo_rl/algorithms/advantage_estimator.py]: https://github.com/NVIDIA-NeMo/RL/pull/2069/changes/77af8bfc4428ba0d4aa6d12dd9ffeeb288ed8531 and also for below comments.

File: nemo_rl/environments/math_environment.py (hunk 10 above)
Comment (yuki-97) [reply nemo_rl/environments/math_environment.py:328]: https://github.com/NVIDIA-NeMo/RL/pull/2069/changes/e23fa7e6d47715955111e74da0b7a1b2cead4d5e and also for other two comments of adding issue.

File: tests/unit/algorithms/test_grpo.py (hunk 4 above)
Comment (yuki-97) [reply tests/unit/algorithms/test_grpo.py:1659]: https://github.com/NVIDIA-NeMo/RL/pull/2069/changes/198c991ea4fff51720d3980663dd038ee5c8a52a

File: nemo_rl/environments/interfaces.py (hunk 7 above)
Comment (nbasyl) [reply nemo_rl/environments/interfaces.py]: Addressed

File: docs/guides/grpo.md [hunk 11]
```diff
@@ -511,6 +511,40 @@ This expectation is estimated using the rollouts in each global training batch a
...
+To enable GDPO for multi-reward RL training, simply set:
+```
+grpo:
+  adv_estimator:
+    name: "gdpo"
+    normalize_rewards: true
+    use_leave_one_out_baseline: false
+```
+Note that this method only has an effect when training involve more than one reward function.
+
```
Comment (yuki-97) [inline docs/guides/grpo.md:564]: nit: could we remove some of the empty lines?

File: README.md [hunk 12]
```diff
@@ -10,6 +10,11 @@
 </div>
 
 ## 📣 News
+* [13/03/2026] GDPO Support
```
Comment (yuki-97) [inline README.md]: nit:
```suggestion
* [12/03/2026] GDPO Support
```

File: nemo_rl/experience/rollouts.py (hunk 9 above)
Comment (yuki-97) [reply nemo_rl/experience/rollouts.py:571]: tracking at https://github.com/NVIDIA-NeMo/RL/issues/2107 and added explanations in the doc.
